waiter/src/waiter/auth/kerberos.clj

```diff
@@ -101,16 +102,19 @@
                          :status 403
                          :user run-as-user}))))))
 
-(defrecord KerberosAuthenticator [password]
+(defrecord KerberosAuthenticator [^ExecutorService thread-pool password]
   auth/Authenticator
   (wrap-auth-handler [_ request-handler]
-    (spnego/require-gss request-handler password)))
+    (spnego/require-gss request-handler thread-pool password)))
 
 (defn kerberos-authenticator
   "Factory function for creating Kerberos authenticator middleware"
-  [{:keys [password]}]
-  {:pre [(not-empty password)]}
-  (->KerberosAuthenticator password))
+  [{:keys [concurrency-level password]}]
+  {:pre [(not-empty password)
+         (integer? concurrency-level)
+         (pos? concurrency-level)]}
+  (let [thread-pool (Executors/newFixedThreadPool concurrency-level)]
```

Comment:
I don't like creating a possibly idle pool of 30+ extra threads. This should be self-scaling, up to the limit you set. I'd also rename the config value max-concurrency-level, not concurrency-level. 

What about 'new ThreadPoolExecutor' directly to construct it, that way you can make the pool auto-scale. Set a core pool size of, say, 1 thread, scale to max-concurrency-level maximum, with a 5 minute idle thread timeout. You can also put in a ThreadFactory and name the threads as they're created! :) That'll make stack traces and monitoring work nicer.

Reply:
Thanks for the suggestion. Agreed and implemented (except the custom thread factory).